src/lsm/grid.zig

```diff
@@ -507,73 +460,85 @@ pub fn GridType(comptime Storage: type) type {
-                    );
-                } else {
-                    unreachable;
-                }
+            if (!header.valid_checksum_body(block[@sizeOf(vsr.Header)..header.size])) {
+                log.err("invalid checksum body at address {}", .{address});
+                return false;
+            }
 
-                // IOP reads that fail checksum validation get punted to a recovery queue.
-                // TODO: Have the replica do something with the pending reads here.
-                while (iop.reads.pop()) |read| {
-                    iop.grid.read_recovery_queue.push(read);
-                }
+            if (header.checksum != checksum) {
+                log.err(
+                    "expected address={} checksum={} block_type={}, " ++
+                        "found address={} checksum={} block_type={}",
+                    .{
+                        address,
+                        checksum,
+                        block_type,
+                        header.op,
+                        header.checksum,
+                        @enumToInt(header.operation),
+                    },
+                );
+                return false;
             }
 
-            grid.read_iops.release(iop);
+            assert(header.op == address);
+            assert(header.operation == block_type.operation());
+            return true;
+        }
 
-            // Always iterate through the full list of pending reads instead of just one to ensure
-            // that those serviced from the cache don't prevent others waiting for an IOP from
-            // seeing the IOP that was just released.
-            var copy = grid.read_queue;
-            grid.read_queue = .{};
-            while (copy.pop()) |read| {
-                assert(read.address != address);
-                grid.start_read(read);
+        fn read_block_resolve(grid: *Grid, read: *Grid.Read, block: BlockPtrConst) void {
+            {
+                // Guard to make sure the cache cannot be updated by pending_read.callbacks()
+                assert(!grid.read_resolving);
+                grid.read_resolving = true;
+                defer grid.read_resolving = false;
+
+                // Resolve all reads queued to the address with the block.
+                // Callbacks may queue more to read.resolves so it must drain any new/existing pendings.
+                while (read.resolves.pop()) |pending| {
+                    const pending_read = @fieldParentPtr(Read, "pending", pending);
+                    pending_read.callback(pending_read, block);
+                }
             }
+            
+            // Remove the "root" read so that the address is no longer actively reading / locked.
+            // Then invoke the callback with the cache block (which is only valid until the the
+            // next Grid.read_block/Grid.write_block which may be done inside callback).
+            grid.read_queue.remove(read);
+            read.callback(read, block);
```

Comment:
Should the `read_resolving` guard cover this callback too?

Reply:
I dont think it's necessary as, by then, the block is safe to be invalidated by the callback. Naming-wise, it's also not accessing `read.resolves` anymore.

Reply:
Suppose read.callback is something like:

```
fn on_read(block) {
  const self = @fieldParentPtr(...);
  // Pipeline the next read
  if (self.next_address()) |address| grid.read_block(on_read, address);
  // `block` may now be invalidated
  self.merge(block);
}
```

I don't think it's safe to invalidate the block until the end of the callback.

Reply:
It has to support invalidation as this is currently an expected flow:
```
TableIterator.tick()
Grid.read_block()
it.index_block_callback(block)
Compaction.iterator_b_info_callback(block)
Grid.release(block)
Grid.assert_not_reading(block)
```

The comment above is meant to highlight this use-case, although it may need better wording:
> (which is only valid until the the next Grid.read_block/Grid.write_block which may be done inside callback)

In the read.callback example, the call to `grid.read_block(on_read, address)` invalidates the `block` before merge() if `block` comes from the same address.

Reply:
The invariant I'm trying to ensure is that during `on_read(block)` it's not possible for any call to `grid` to change the value of `block`, because that kind of non-local aliasing is incredibly hard to catch when reading the code. (I think the current code does not do this, but the assert is nice to have for future editors of the code.)

The flow you gave would be fine even with the extended assert because write_block doesn't update the cache synchronously and release doesn't mutate the value of the cache entry, just reduces its count to 1.

Reply:
To rephrase the point: the `on_read` I gave above is safe with the grid impl in this pr, I believe, and if we make that invariant explicit then we can pipeline compaction without having to always copy blocks.

Reply:
And we might as well, because even if the invariant was

> (which is only valid until the the next Grid.read_block/Grid.write_block which may be done inside callback)

then we would only be able to take advantage of it in the root read - the resolves still have to be valid until the end of the callback.

Reply:
Oh, is the problem that you need to remove from `read_queue` before `read.callback`?

Reply:
> The flow you gave would be fine even with the extended assert because write_block doesn't update the cache synchronously and release doesn't mutate the value of the cache entry [..] is the problem you're thinking about that you need to remove from read_queue before read.callback

Oh ur right. I keep forgetting that `start_read/start_write` no longer update the cache synchronously themselves anymore. Makes sense now that we can extend the `read_resolving` guard to include to root callback too.

Reply:
Yeah, `on_next_tick` makes a lot of things easier to reason about. I'm excited to get this in :)

Once this is merged I'll follow up with the block node pool, which should also make it a bit easier to think about cache safety.

Reply:
:metal: